### PR TITLE
Support Array with the `json` option

### LINKED ISSUE #334: Support posting Arrays with json option
Eventhough it was mentioned in #174:

> I'm ok with also supporting an array

Arrays are not supported with the `json` option (because of `isPlainObject`) so this fails:
```js
got.post('example.com', {
  json: true,
  body: [1, 2, 3]
})
```
Could you use something like [lodash's `isObjectLike`](https://lodash.com/docs#isObjectLike) instead of `isPlainObject` or test both `isPlainObject || isArray`?

### PR DESCRIPTION
**Pull request by DerTieran**
Allow to post Arrays with the `json` option.
This was discussed in [#174](https://github.com/sindresorhus/got/issues/174#issuecomment-299534498) but not implemented.

> I'm ok with also supporting an array

Fixes #334

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -315,8 +315,10 @@ function normalizeArguments(url, opts) {
 			throw new TypeError('options.body must be a ReadableStream, string, Buffer or plain Object');
 		}
 
-		if ((opts.form || opts.json) && !isPlainObj(body)) {
-			throw new TypeError('options.body must be a plain Object when options.form or options.json is used');
+		if (opts.form && !isPlainObj(body)) {
+			throw new TypeError('options.body must be a plain Object when options.form is used');
+		} else if (opts.json && !(isPlainObj(body) || Array.isArray(body))) {
```
Comment (kevva) [inline index.js]: Don't need a `else if` here, just do a new `if` block.

Comment (DerTieran) [inline index.js]: True, changed it.
*(Hope you guys are okay with amending the last commit)*

File: index.js [hunk 2]
```diff
@@ -315,8 +315,12 @@ function normalizeArguments(url, opts) {
 			throw new TypeError('options.body must be a ReadableStream, string, Buffer or plain Object');
 		}
 
-		if ((opts.form || opts.json) && !isPlainObj(body)) {
```
Comment (AlexTes) [inline index.js]: I'd prefer if we kept this joint, it's simpler. Let's discuss it over in the issue.

File: test/post.js [hunk 3]
```diff
@@ -121,7 +129,7 @@ test('content-type header is not overriden when object in options.body', async t
 	t.is(headers['content-type'], 'doge');
 });
 
-test('throws when json body is not a plain object', async t => {
+test('throws when json body is not a plain object or array', async t => {
```
Comment (AlexTes) [inline test/post.js]: If we do choose to split the argument validation for `json` from `form` we should also have two tests to make sure both work the way we expect.

File: index.js [hunk 4]
```diff
@@ -315,17 +315,17 @@ function normalizeArguments(url, opts) {
...
 
 		if (isStream(body) && typeof body.getBoundary === 'function') {
 			// Special case for https://github.com/form-data/form-data
 			headers['content-type'] = headers['content-type'] || `multipart/form-data; boundary=${body.getBoundary()}`;
-		} else if (opts.form && isPlainObj(body)) {
+		} else if (opts.form && (isPlainObj(body) || Array.isArray(body))) {
 			headers['content-type'] = headers['content-type'] || 'application/x-www-form-urlencoded';
 			opts.body = querystring.stringify(body);
-		} else if (opts.json && isPlainObj(body)) {
+		} else if (opts.json && (isPlainObj(body) || Array.isArray(body))) {
```
Comment (sindresorhus) [inline index.js]: You're repeating `(isPlainObj(body) || Array.isArray(body))` 3 times. That should be DRY'd up.

**Comment by sindresorhus** [comment]
The docs needs to be updated to reflect this change.

File: index.js (hunk 4 above)
Comment (DerTieran) [inline index.js]: Sure something like `isPlainObjOrArray`?

Comment (sindresorhus) [inline index.js]: `isValidBody` or something. Better to describe the intent than the contents.

**Pull request merged**